[Bug] EKS 1.28 with Karpenter v0.33.1 not working

<!--
Thank you for opening this bug report! You are helping make eksctl a better project :)
In order to help us process this issue faster, please provide as much detail as possible.

When providing command examples, error output or code samples, please use code blocks.
For verbose output, for example logs, please either create a gist or use html details tags.
-->

<!--
Before you start writing:
1. Please search through existing issues to see whether your problem has already been reported.
2. Download the latest version of eksctl to see whether the problem has already been fixed.

For more information on how to open a bug report in eksctl, please refer to our issue submission guide.
https://github.com/eksctl-io/eksctl/blob/main/CONTRIBUTING.md#opening-issues
-->

### What were you trying to accomplish?
<!-- Information about the command you ran and what you expected to happen. -->
trying to create cluster 1.28 with karpenter v0.33.1

### What happened?
<!-- A description of actual behavior (with error messages). -->

cluster created but karpenter hanged

```
2024-01-16 07:49:37 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-karpenter"
2024-01-16 07:50:07 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-karpenter"
2024-01-16 07:50:46 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-karpenter"
2024-01-16 07:51:26 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-karpenter"
2024-01-16 07:52:27 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-karpenter"
2024-01-16 07:52:27 [ℹ]  1 task: { create IAM role for serviceaccount "karpenter/karpenter" }
2024-01-16 07:52:27 [ℹ]  1 task: { create IAM role for serviceaccount "karpenter/karpenter" }
2024-01-16 07:52:27 [ℹ]  building iamserviceaccount stack "eksctl-eks-kptv3-addon-iamserviceaccount-karpenter-karpenter"
2024-01-16 07:52:27 [ℹ]  deploying stack "eksctl-eks-kptv3-addon-iamserviceaccount-karpenter-karpenter"
2024-01-16 07:52:27 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-addon-iamserviceaccount-karpenter-karpenter"
2024-01-16 07:52:57 [ℹ]  waiting for CloudFormation stack "eksctl-eks-kptv3-addon-iamserviceaccount-karpenter-karpenter"
2024-01-16 07:52:57 [ℹ]  adding identity "arn:aws:iam::xxxxxxxxxx:role/eksctl-KarpenterNodeRole-eks-kptv3" to auth ConfigMap
2024-01-16 07:52:57 [ℹ]  adding Karpenter to cluster eks-kptv3
E0116 07:52:59.856684    3205 memcache.go:206] couldn't get resource list for karpenter.k8s.aws/v1beta1: the server could not find the requested resource
E0116 07:52:59.942104    3205 memcache.go:206] couldn't get resource list for karpenter.sh/v1beta1: the server could not find the requested resource
Error: failed to install Karpenter: failed to install Karpenter chart: failed to install chart: timed out waiting for the condition
```


### How to reproduce it?
<!--
Include ALL the steps to reproduce the bug.

If using a config file, include it here, removing any sensitive information!
-->
```shell
eksctl create cluster -f ./cluster.yaml
```

```yaml
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: eks-kptv3
  region: ap-northeast-1
  version: "1.28"
  tags:
    karpenter.sh/discovery: eks-kptv3 # here, it is set to the cluster name
iam:
  withOIDC: true # required

karpenter:
  version: 'v0.33.1' # Exact version must be specified
  createServiceAccount: true # default is false
  withSpotInterruptionQueue: true

vpc:
  cidr: 10.10.0.0/16
  hostnameType: resource-name
  # disable public access to endpoint and only allow private access
  clusterEndpoints:
    publicAccess: true

managedNodeGroups:
  - name: ng-apps
    instanceType: t3.medium
    minSize: 2
    maxSize: 4
    desiredCapacity: 2
    volumeSize: 20
```
**Logs**
<!--
Include the output of the command line when running eksctl. If possible, eksctl should be run with debug logs. For example:
`eksctl get clusters -v 4`
Make sure you redact any sensitive information before posting.
If the output is long, please consider a Gist, or an html details tag.
-->


**Anything else we need to know?**
<!--
What OS are you using?
Are you using a downloaded binary or did you compile eksctl?
What type of AWS credentials are you using (i.e. default/named profile, MFA)? - please don't include actual credentials though!
-->

**Versions**
<!--
Please paste in the output of this command:
-->
```
$ eksctl info
eksctl version: 0.167.0
kubectl version: v1.29.0
OS: linux

$helm version
version.BuildInfo{Version:"v3.13.3", GitCommit:"c8b948945e52abba22ff885446a1486cb5fd3474", GitTreeState:"clean", GoVersion:"go1.20.11"}
```

I have reproduced this problem: 

```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: cluster-with-karpenter-1
  region: eu-west-1
  version: '1.28'
  tags:
    karpenter.sh/discovery: cluster-with-karpenter-1 # here, it is set to the cluster name
iam:
  withOIDC: true # required

karpenter:
  version: 'v0.33.1' # Exact version must be specified

managedNodeGroups:
  - name: managed-ng-1
    minSize: 2
    maxSize: 2
    desiredCapacity: 2
```

The helm chart has failed to install

```
$ helm -n karpenter history karpenter
REVISION        UPDATED                         STATUS  CHART                   APP VERSION     DESCRIPTION                                                    
1               Wed Jan 17 12:11:51 2024        failed  karpenter-v0.33.1       0.33.1          Release "karpenter" failed: timed out waiting for the condition
```

Pods are crashlooping 

```
$ kubectl get pods -n karpenter
NAME                        READY   STATUS             RESTARTS         AGE
karpenter-b5475c9cc-jsgns   0/1     CrashLoopBackOff   19 (2m41s ago)   74m
karpenter-b5475c9cc-l9jrr   0/1     CrashLoopBackOff   19 (2m32s ago)   74m
```

Log

```
$ kubectl logs karpenter-b5475c9cc-jsgns -n karpenter
panic: validating options, missing field, cluster-name

goroutine 1 [running]:
github.com/samber/lo.must({0x25cbe60, 0xc000606c80}, {0x0, 0x0, 0x0})
        github.com/samber/lo@v1.39.0/errors.go:53 +0x1e9
github.com/samber/lo.Must0(...)
        github.com/samber/lo@v1.39.0/errors.go:72
sigs.k8s.io/karpenter/pkg/operator/injection.WithOptionsOrDie({0x30c0598, 0xc00027b9e0}, {0xc0005ac9c0, 0x2, 0x2333e00?})
        sigs.k8s.io/karpenter@v0.33.1/pkg/operator/injection/injection.go:51 +0x138
sigs.k8s.io/karpenter/pkg/operator.NewOperator()
        sigs.k8s.io/karpenter@v0.33.1/pkg/operator/operator.go:84 +0xb7
main.main()
        github.com/aws/karpenter/cmd/controller/main.go:33 +0x25
```

Hope this helps

--- a/pkg/karpenter/karpenter.go
+++ b/pkg/karpenter/karpenter.go
@@ -9,6 +9,7 @@ import (
 
 	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
 	"github.com/weaveworks/eksctl/pkg/karpenter/providers"
+	"github.com/weaveworks/eksctl/pkg/utils"
 )
 
 const (
@@ -78,14 +79,20 @@ func (k *Installer) Install(ctx context.Context, serviceAccountRoleARN string, i
 			defaultInstanceProfile: instanceProfileName,
 		},
 		settings: map[string]interface{}{
-			aws: map[string]interface{}{
-				defaultInstanceProfile: instanceProfileName,
-				clusterName:            k.ClusterConfig.Metadata.Name,
-				clusterEndpoint:        k.ClusterConfig.Status.Endpoint,
-				interruptionQueueName:  k.ClusterConfig.Metadata.Name,
-			},
+			defaultInstanceProfile: instanceProfileName,
+			clusterName:            k.ClusterConfig.Metadata.Name,
+			clusterEndpoint:        k.ClusterConfig.Status.Endpoint,
+			interruptionQueueName:  k.ClusterConfig.Metadata.Name,
 		},
 		serviceAccount: serviceAccountMap,
+	}
+
+	version := k.ClusterConfig.Karpenter.Version
+	compareVersions, err := utils.CompareVersions(version, "0.33.0")
+	if err == nil && compareVersions < 0 {
+		values[settings] = map[string]interface{}{
+			aws: values[settings],
+		}
 	}
 
 	registryClient, err := registry.NewClient(
@@ -101,7 +108,7 @@ func (k *Installer) Install(ctx context.Context, serviceAccountRoleARN string, i
 		Namespace:       DefaultNamespace,
 		ReleaseName:     releaseName,
 		Values:          values,
-		Version:         k.ClusterConfig.Karpenter.Version,
+		Version:         version,
 		RegistryClient:  registryClient,
 	}
 

--- a/pkg/karpenter/karpenter_test.go
+++ b/pkg/karpenter/karpenter_test.go
@@ -81,6 +81,21 @@ var _ = Describe("Install", func() {
 			}))
 		})
 
+		It("installs karpenter with expanded settings.aws values for version greater or equal to v0.33.0", func() {
+			installerUnderTest.Options.ClusterConfig.Karpenter.Version = "0.33.0"
+			Expect(installerUnderTest.Install(context.Background(), "dummy", "dummy")).To(Succeed())
+			_, opts := fakeHelmInstaller.InstallChartArgsForCall(0)
+			values := map[string]interface{}{
+				settings: map[string]interface{}{
+					defaultInstanceProfile: "dummy",
+					clusterName:            cfg.Metadata.Name,
+					clusterEndpoint:        cfg.Status.Endpoint,
+					interruptionQueueName:  cfg.Metadata.Name,
+				},
+			}
+			Expect(opts.Values[settings]).To(Equal(values[settings]))
+		})
+
 		When("install chart fails", func() {
 
 			BeforeEach(func() {

